File: frontend/src/types/draws.ts
export type LotteryType = "ssq" | "dlt";

export type DrawPublic = {
  id: string;
  lottery_type: LotteryType;
  issue: string;
  draw_date: string;
  primary_numbers: number[];
  secondary_numbers: number[];
  source: string;
  source_checksum: string | null;
  created_at: string;
  updated_at: string;
};

export type DrawListData = {
  items: DrawPublic[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
};

export type LatestDrawsData = {
  ssq: DrawPublic | null;
  dlt: DrawPublic | null;
};

export type SyncAcceptedData = {
  job_id: string;
  run_id: string;
  lottery_type: LotteryType;
  mode: "incremental" | "full";
  status: string;
};

export type IngestionRunPublic = {
  id: string;
  job_id: string | null;
  source_name: string;
  lottery_type: string;
  mode: string;
  status: string;
  started_at: string | null;
  finished_at: string | null;
  pages_processed: number;
  records_seen: number;
  inserted_count: number;
  updated_count: number;
  skipped_count: number;
  error_count: number;
  cursor: string | null;
  error_summary: string | null;
};

export type IngestionRunListData = {
  items: IngestionRunPublic[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
};
export type ImportPreviewRow = {
  row_number: number;
  lottery_type: string | null;
  issue: string | null;
  draw_date: string | null;
  primary_numbers: number[] | null;
  secondary_numbers: number[] | null;
  valid: boolean;
  errors: string[];
};

export type ImportPreviewData = {
  total_rows: number;
  valid_rows: number;
  invalid_rows: number;
  rows: ImportPreviewRow[];
};

export type ImportCommitData = {
  inserted_count: number;
  updated_count: number;
  skipped_count: number;
  error_count: number;
};

File: backend/app/schemas/draws.py
"""Draw and ingestion schemas."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

LotteryType = Literal["ssq", "dlt"]
SyncMode = Literal["incremental", "full"]


class DrawPublic(BaseModel):
    id: UUID
    lottery_type: LotteryType
    issue: str
    draw_date: date
    primary_numbers: list[int]
    secondary_numbers: list[int]
    source: str
    source_checksum: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DrawDetail(DrawPublic):
    raw_payload: dict[str, Any] | None = None


class DrawListData(BaseModel):
    items: list[DrawPublic]
    page: int
    page_size: int
    total: int
    total_pages: int


class LatestDrawsData(BaseModel):
    ssq: DrawPublic | None = None
    dlt: DrawPublic | None = None


class SyncRequest(BaseModel):
    lottery_type: LotteryType
    mode: SyncMode = "incremental"
    page_size: int = Field(default=30, ge=10, le=100)


class SyncAcceptedData(BaseModel):
    job_id: UUID
    run_id: UUID
    lottery_type: LotteryType
    mode: SyncMode
    status: str


class IngestionRunPublic(BaseModel):
    id: UUID
    job_id: UUID | None
    source_name: str
    lottery_type: str
    mode: str
    status: str
    started_at: datetime | None
    finished_at: datetime | None
    pages_processed: int
    records_seen: int
    inserted_count: int
    updated_count: int
    skipped_count: int
    error_count: int
    cursor: str | None
    error_summary: str | None

    model_config = {"from_attributes": True}


class IngestionRunListData(BaseModel):
    items: list[IngestionRunPublic]
    page: int
    page_size: int
    total: int
    total_pages: int


class ImportPreviewRow(BaseModel):
    row_number: int
    lottery_type: str | None = None
    issue: str | None = None
    draw_date: str | None = None
    primary_numbers: list[int] | None = None
    secondary_numbers: list[int] | None = None
    valid: bool
    errors: list[str] = Field(default_factory=list)


class ImportPreviewData(BaseModel):
    total_rows: int
    valid_rows: int
    invalid_rows: int
    rows: list[ImportPreviewRow]


class ImportCommitRequest(BaseModel):
    rows: list[dict[str, Any]]


class ImportCommitData(BaseModel):
    inserted_count: int
    updated_count: int
    skipped_count: int
    error_count: int
